### PR TITLE
KeySignature - port to TypeScript

### PR DESCRIPTION
**Pull request by ksraj123**
Reference #18 
Ports `keysignature.py` to typescript.

### CODE REVIEW

File: src/.prototype/musicutils/KeySignature.ts [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+                    }
+                }
+            }
+        } else {
+            // Convert to preferred accidental.
+            if (this.key.indexOf('#') !== -1) {
+                scale.forEach((note: string, i: number) => {
+                    if (note.indexOf('b') !== 0) {
+                        if (note in MusicUtils.EQUIVALENT_SHARPS) {
+                            scale[i] = MusicUtils.EQUIVALENT_SHARPS[note];
```
Comment (ksraj123) [inline src/.prototype/musicutils/KeySignature.ts:232]: What to you think about encapsulating exports from modules like util modules that export a lot of things into a single entity? I think it makes the code a little more readable as it immediately clear where something comes from.

**Review by ksraj123 — COMMENTED** [review]
@meganindya This PR ports only a part of the file, so that its easier to go through line by line. Please have a look at the ported and original code and share your insights. I have also highlighted some points I was curious about. Thanks

File: src/.prototype/musicutils/abstract/AKeySignature.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
+export abstract class AKeySignature {
```
[the remark this answers is not in the document]
Comment (meganindya) [reply src/.prototype/musicutils/abstract/AKeySignature.ts]: Exactly. If you see some of my files in `syntax-core`, I've tried to use interfaces, abstract classes and methods, etc. When we're using TS, we better use the features thoroughly. Yet again, not really very useful for single class cases. Could use interfaces though &mdash; would help an onlooker identify the members of the class all at once.

File: src/.prototype/musicutils/KeySignature.ts (hunk 1 above)
Comment (meganindya) [reply src/.prototype/musicutils/KeySignature.ts:232]: Surely there's better organization possible like you suggest. I'll take a dab at the classes and get back later today.

Comment (meganindya) [reply src/.prototype/musicutils/KeySignature.ts:232]: Integrating all the exports as a JSON at the end of the file for long files seems like a better approach to me. Kind of gives the interface-like feel.

**Comment by ksraj123** [comment]
Entire file ported, all the test cases are passing. Thanks

Screenshot 2021-03-22 at 9 10 54 AM

File: src/.prototype/musicutils/ts/keysignature.ts [hunk 3]
```diff
@@ -0,0 +1,1500 @@
...
+    public get numberOfSemitones(): number {
+        return this._numberOfSemitones;
+    }
+
+    /**
+     * Custom note names defined by user
+     * @param customNames - A list of custom names
+     * @returns
+     */
+    public set customNoteNames(customNames: string[]) {
```
Comment (meganindya) [inline src/.prototype/musicutils/ts/keysignature.ts:572]: @walterbender don't we need a unique check too?

**Comment by meganindya** [comment]
This is taking a while. I'm making a few changes on top. Converting to draft for now.

File: src/.prototype/musicutils/ts/keysignature.ts [hunk 4]
```diff
@@ -0,0 +1,1640 @@
...
+        'shahnaz': 'd',
+        'maqam mustar': 'eb',
+        'maqam jiharkah': 'f',
+        'shadd araban': 'g',
+        'suzidil': 'a',
+        'ajam': 'bb',
+        'ajam maqam': 'bb'
+    };
+
+    static readonly MODE_MAPPER: { [key: string]: string } = {
```
Comment (meganindya) [inline src/.prototype/musicutils/ts/keysignature.ts:122]: docs

File: src/.prototype/musicutils/ts/keysignature.ts [hunk 5]
```diff
@@ -0,0 +1,1640 @@
...
+    ];
+
+    key: string;
+    mode: string;
+    scale: string[];
+    halfSteps: number[];
+    keySignature: string;
+    genericScale: string[];
+    scalarModeNumbers: string[];
+    eastIndianSolfegeNotes: string[];
```
Comment (meganindya) [inline src/.prototype/musicutils/ts/keysignature.ts:205]: Which of these should we expect to be directly accessed through the instance, and which are private?

Also need docs.

File: src/.prototype/musicutils/ts/keysignature.ts [hunk 6]
```diff
@@ -0,0 +1,1640 @@
...
+
+    private _scale: Scale;
+    private _noteNames: string[];
+    private _fixedSolfege: boolean;
+    private _solfegeNotes: string[];
+    private _customNoteNames: string[];
+    private _numberOfSemitones: number;
```
Comment (meganindya) [inline src/.prototype/musicutils/ts/keysignature.ts:212]: docs

File: src/.prototype/musicutils/ts/keysignature.ts [hunk 7]
```diff
@@ -0,0 +1,1640 @@
...
+     * Converts from a letter name used by 12-semitone temperaments to a generic note name as
+     * defined by the temperament.
+     *
+     * @remarks
+     * Only for temperaments with 12 semitones.
+     *
+     * @param pitchName - name of pitch.
+     * @returns
+     */
+    public convertToGenericNoteName(pitchName: string): [string, number] {
```
Comment (meganindya) [inline src/.prototype/musicutils/ts/keysignature.ts:677]: what is the string variable and what is the number variable

File: src/.prototype/musicutils/ts/keysignature.ts [hunk 8]
```diff
@@ -0,0 +1,1640 @@
...
+        }
+
+        console.debug(`Note name ${noteName} not found.`);
+        return [noteName, -1];
+    }
+
+    /**
+     * @param noteName
+     * @param targetList
+     * @returns
```
Comment (meganindya) [inline src/.prototype/musicutils/ts/keysignature.ts:869]: docs

File: src/.prototype/musicutils/ts/keysignature.ts [hunk 9]
```diff
@@ -0,0 +1,1640 @@
...
+    /**
+     * Convert from a generic note name as defined by the temperament to a letter name used by
+     * 12-semitone temperaments.
+     *
+     * @remarks
+     * Only for temperaments with 12 semitones.
+     *
+     * @param noteName
+     * @param preferSharps
+     * @returns
```
Comment (meganindya) [inline src/.prototype/musicutils/ts/keysignature.ts:827]: docs

File: src/.prototype/musicutils/ts/keysignature.ts [hunk 10]
```diff
@@ -0,0 +1,1640 @@
...
+        console.debug(`Note name ${noteName} not found.`);
+        return [noteName, -1];
+    }
+
+    /**
+     * @param noteName
+     * @param sharpScale
+     * @param flatScale
+     * @param preferSharps
+     * @returns
```
Comment (meganindya) [inline src/.prototype/musicutils/ts/keysignature.ts:915]: docs

File: src/.prototype/musicutils/ts/keysignature.ts [hunk 11]
```diff
@@ -0,0 +1,1640 @@
...
+    /**
+     * Converts from a generic note name as defined by the temperament to a solfege note used by
+     * 12-semitone temperaments.
+     *
+     * @remarks
+     * Only for temperaments with 12 semitones.
+     *
+     * @param noteName
+     * @param preferSharps
+     * @returns
```
Comment (meganindya) [inline src/.prototype/musicutils/ts/keysignature.ts:960]: docs

File: src/.prototype/musicutils/ts/keysignature.ts [hunk 12]
```diff
@@ -0,0 +1,1640 @@
...
+    /**
+     * Converts from a generic note name as defined by the temperament to an East Indian solfege
+     * note used by 12-semitone temperaments.
+     *
+     * @remarks
+     * Only for temperaments with 12 semitones.
+     *
+     * @param noteName
+     * @param preferSharps
+     * @returns
```
Comment (meganindya) [inline src/.prototype/musicutils/ts/keysignature.ts:980]: docs

File: src/.prototype/musicutils/ts/keysignature.ts [hunk 13]
```diff
@@ -0,0 +1,1640 @@
...
+
+    /**
+     * Converts from a generic note name as defined by the temperament to a custom_noteName used by
+     * 12-semitone temperaments.
+     *
+     * @remarks
+     * Only for temperaments with 12 semitones.
+     *
+     * @param noteName
+     * @returns
```
Comment (meganindya) [inline src/.prototype/musicutils/ts/keysignature.ts:1019]: docs

File: src/.prototype/musicutils/ts/keysignature.ts [hunk 14]
```diff
@@ -0,0 +1,1640 @@
...
+     * Given a starting pitch, add a semitone transform and return the resultant pitch (and any
+     * change in octave).
+     *
+     * @param startingPitch - The starting pitch specified as a pitch letter, e.g., c#, fb.
+     * @param numberOfHalfSteps - Half steps are steps in the notes defined by the temperament.
+     * @returns
+     * An array (3-tuple) of
+     *  - The pitch that is the number of half steps from the starting pitch.
+     *  - The relative change in octave between the starting pitch and the new pitch.
+     *  - Error code.
```
Comment (meganindya) [inline src/.prototype/musicutils/ts/keysignature.ts:1101]: error values docs

File: src/.prototype/musicutils/ts/keysignature.ts [hunk 15]
```diff
@@ -0,0 +1,1640 @@
...
+     * @returns
+     * An array (4-tuple) of
+     *  - The closest pitch to the target pitch in the scale.
+     *  - The scalar index value of the closest pitch in the scale (If the target is midway between
+     *      two scalar pitches, the lower pitch is returned.)
+     *  - The distance in semitones (half steps) from the target pitch to the scalar pitch (If the
+     *      target is higher than the scalar pitch, then distance > 0. If the target is lower than
+     *      the scalar pitch then distance < 0. If the target matches a scale pitch, then distance
+     *      is 0.)
+     *  - Error code (0 means success).
```
Comment (meganindya) [inline src/.prototype/musicutils/ts/keysignature.ts:1544]: error values docs

**Review by meganindya — COMMENTED** [review]
@walterbender need some inputs.

Documentation required at some more places. I've marked the areas with a 'docs' comment.

On another note, there are 16 `print` statements in the Python code. I was thinking if we should use `throw`-`try`-`catch` error handling for those cases instead?

File: src/.prototype/musicutils/ts/keysignature.ts [hunk 16]
```diff
@@ -0,0 +1,1640 @@
...
+     * @remarks
+     * Names should not end with b or x or they will cause collisions with the flat (b) and
+     * doublesharp (x) accidentals.
+     *
+     * @param customNames - A list of custom names.
+     *
+     * @throws {InvalidArgumentError}
+     * Thrown if exactly same number of custom names aren't supplied for the mode length.
+     */
+    public set customNoteNames(customNames: string[]) {
```
Comment (meganindya) [inline src/.prototype/musicutils/ts/keysignature.ts:598]: shouldn't we require a unique check too?

**Review by meganindya — COMMENTED** [review]
Took me forever to finish this up. Great job on your part though.

**Pull request merged**